msticpy/init/mp_user_session.py

```diff
@@ -0,0 +1,178 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+MSTICPy User Session loader.
+
+This loads user-specific configuration settings from a YAML file.
+
+Example YAML file:
+
+.. code-block:: yaml
+
+    QueryProviders:
+        qry_prov_sent:
+            DataEnvironment: MSSentinel
+            InitArgs:
+                debug: True
+            Connect: True
+            ConnectArgs:
+                workspace: MySoc
+                auth_methods: ['cli', 'device_code']
+        qry_prov_md:
+            DataEnvironment: M365D
+        qry_kusto_mde:
+            DataEnvironment: Kusto
+            Connect: True
+            ConnectArgs:
+                cluster: MDEData
+        qry_kusto_mstic:
+            DataEnvironment: Kusto
+            Connect: True
+            ConnectArgs:
+                cluster: MSTIC
+    Components:
+        mssentinel:
+            Module: msticpy.context.azure
+            Class: MicrosoftSentinel
+            InitArgs:
+            Connect: True
+            ConnectArgs:
+                workspace: CyberSecuritySoc
+                auth_methods: ['cli', 'device_code']
+
+Example usage:
+
+.. code-block:: python
+
+    import msticpy as mp
+    mp.init_notebook()
+    mp.mp_user_session.load_user_config()
+
+"""
+
+from __future__ import annotations
+
+import importlib
+import sys
+from pathlib import Path
+from typing import Any
+
+import yaml
+from IPython import get_ipython
+
+from .._version import VERSION
+from ..data.core.data_providers import QueryProvider
+
+__version__ = VERSION
+__author__ = "Ian Hellen"
+
+
+def load_user_session(
+    session_file: str | Path | None = None, namespace: dict[str, Any] | None = None
+):
+    """
+    Load user session configuration.
+
+    Parameters
+    ----------
+    session_file : str, Path, optional
+        Path to the user configuration file, by default None.
+        If no file is provided, the function will look in the current directory
+        for an `mp_user_session.yaml` file.
+    namespace : dict[str, Any], optional
+        Namespace to load the configuration into, by default None.
+        If no namespace is provided, the configuration will be loaded into the
+        global namespace of the caller (usually the notebook).
+
+    """
+    if session_file is None:
+        session_file = Path.cwd() / "mp_user_session.yaml"
+    session_file = Path(session_file)
+    if not session_file.exists():
+        print(f"Session file {session_file} not found.")
+        return
+
+    if namespace is None:
+        ipython = get_ipython()
+        if ipython is None:
+            # pylint: disable=protected-access
+            namespace = sys._getframe(1).f_locals
```

Comment:
TIL.... neat!

Reply:
Learned this from a Python Guru David Beasley many years ago!